### DISCUSSION TITLE
Finding a Candidate List for a Website

### DISCUSSION THREAD

**Discussion by aswilkinson12** [McGill-NLP/weblinx#26] [Q&A] [answered]
I'm working on the Applications track for COMP 545 regarding implementation of Weblinx, and so far I understand the algorithm of how it works and what to feed in, the problem I'm encountering is finding a list of candidates for a website. I can find the dom tree while simulating a website using Playwright, but not how to get the list of candidates in order to feed the information into the model. Is there a function or method that I maybe missed that gets that list of candidates? Thanks!

**Comment by xhluca** [comment, marked as answer]
The function `build_records_for_single_turn` used in `modeling.dmr.processing` shows how to take a `replay` and `turn` object to format the query and list of candidate elements:

https://github.com/McGill-NLP/weblinx/blob/c8986ae5daa35d6e762d67ed891c43bbdfcaa1b6/modeling/dmr/processing.py#L302-L364

You might want to simplify it, since you don't need the negatives in real-time use, and you also don't have a target_uid during inference (only for training/eval), so you might want something like this:

```python
def format_relevant_elements_for_single_turn(
    turn, format_intent_input, uid_key="data-webtasks-id"
) -> List[dict]:
    bboxes_filt = wh.filter_bboxes(
        turn.bboxes,
        viewport_height=turn.viewport_height,
        viewport_width=turn.viewport_width,
    )
    root = lxml.html.fromstring(turn.html)
    root_tree = root.getroottree()
    elements = root.xpath(f"//*[@{uid_key}]")
    elements_filt = [p for p in elements if p.attrib[uid_key] in bboxes_filt]

    records = []

    for elem in elements_filt:
        bbox = turn.bboxes[elem.attrib[uid_key]]
        elem_dict = represent_element_as_dict(elem, bbox, root_tree)
        elem_str = convert_elem_dict_to_str_legacy(elem_dict)

        record = {
            "doc": elem_str,
            "uid": elem.attrib[uid_key],
            "turn_index": turn.index,
            "elem_dict": elem_dict,
        }
        records.append(record)

    return records

# Usage:
from modeling.dmr.processing import build_formatters  # weblinx/modeling/dmr/processing.py
from sentence_transformers.util import cos_sim

uid_key = "what-you-injected-into-elements-on-javascript-side"
format_intent_input, _ = build_formatters()
query = format_turn_for_input(replay, turn, format_intent=format_intent_input)
elements_records = format_relevant_elements_for_single_turn(turn=turn, format_intent_input=format_intent_input, uid_key=uid_key)

# now, use dmr to select candidate elements from the list of relevant formatted elements
# for example (untested, just to illustrate):
docs = [r['doc'] for r in elements_records]
encoded = model.encode(
    [query] + docs, batch_size=batch_size, show_progress_bar=False
)
query_vector, doc_vectors = encoded[0], encoded[1:]
scores = cos_sim(query_vector, doc_vectors).cpu().squeeze().tolist()

for i in range(len(records)):
    records[i]['score'] = scores[i]
```

This code snippet shows you how to do it:

https://github.com/McGill-NLP/weblinx/blob/53b6a234f7f1a458fe0f41c7039a6f69a0693c9d/modeling/dmr/eval.py#L74-L88

Note that you need to construct the `Replay` and `Turn` objects dynamically, which might require some coding effort, since they were originally designed for processing on-disk data that was pre-processed, whereas your goal is to use them for in-memory raw data. The best way to get started is to tinker with the Demonstration/Replay/Turn classes from `weblinx` and understand how they work, then inherit them in your custom classes to use for inference, along the lines of:

```python
import weblinx as wl

class ReplayDynamic(wl.Replay):
    def __init__(self, ..., html=None):
        super().__init__()
        if html is not None:
            self._html_in_memory = html
    
    @cached_property
    def html(self):
        return self._html_in_memory
    # modify any other method/property, or add new methods/properties

class DemonstrationDynamic(wl.Demonstration):
    # also make necessary modifications here

class Turn(wl.Turn):
    # also make necessary modifications here
```

**Reply by aswilkinson12** [reply]
Amazing! Thank you very much for the guidance, it's a great help.

**Reply by xhluca** [reply]
Glad it was helpful! Feel free to share your preferred implementation and we can perhaps add it as a new module, e.g. `weblinx.inference` for any inference-focused code

**Reply by aswilkinson12** [reply]
Hello again! Sorry for the delay before more development, I've been super swamped with work for my other classes. 
If possible, could you give me a bit of guidance / explanation on some of the code you've written here so that I can work with it a bit better? namely, I'm curious about how to work with the "format_relevant_elements_for__single_turn() function that you've written as an example.
Specifically, I'd like a bit of guidance on how to create a "Turn". I currently have [this string](https://pastebin.com/GVtW6Qvd) that we get back, which has a bunch of html portions in the second part. Should I use something like beautifulSoup in order to get the bboxes, or transform it into the proper format? 
Also, for our purposes, how much of the Turn class do you think needs to be "reimplemented"? I was looking through the class itself as it stands within Weblinx and it seems only the bboxes, elements, viewport stuff, html, and index. Do you think off the top of your head any of the other stuff would be useful?
Thank you so much for the guidance! I really want to make this work for inference, as I think it will be pretty cool for the model!
-Anthony

**Reply by xhluca** [reply]
The bounding boxes can only be captured when the webpage is being rendered. it is recommended to do that on the JS side, here's how we did it in JS:

```js
function saveViewportBboxes(eventId) {
    var elementToBoundingBox = {};

    document.querySelectorAll('*').forEach(element => {
        const boundingBox = element.getBoundingClientRect();
        const { innerHeight, innerWidth } = window;

        const isInViewport = !(boundingBox.top > innerHeight || boundingBox.bottom < 0 || boundingBox.left > innerWidth || boundingBox.right < 0);
        const isVisible = boundingBox.width > 1 && boundingBox.height > 1;

        if (isInViewport && isVisible) {
            elementToBoundingBox[element.getAttribute('data-webtasks-id')] = boundingBox;
        }
    });

    // save bboxes to storage
    // make a key for the event id
    const key = "bboxes-" + eventId;
    chrome.storage.local.set({ [key]: { eventId: eventId, bboxes: JSON.stringify(elementToBoundingBox, null, 4) } });
}
```

Note that `'data-webtasks-id'` is injected into each element, you can use whichever unique ID name (`uid_key`) you want, as long as you make sure it is properly added as an attribute to all relevant element in the webpage.

**Reply by aswilkinson12** [reply]
Understood! I'll be sure to pass this on to my partner working with Javascript.
I think currently our thing we need to figure out exactly is how to determine which parts of the webpage are relevant--is that something that you have ideas about how to do? I was thinking about using BeautifulSoup, but I was unsure if there was already a good method of doing so.

**Reply by xhluca** [reply]
it's probably fine to add to anything in viewport. candidate selection is the follow-up stage.

**Reply by georgel** [reply]
Hi, I am working on a similar task to generate new test candidates. What I have tried so far:

1. Using shot-scraper to run Playwright and inject the javascript above 
2. Save html with UID to experiment with DMR and see how well the extraction works

Shot-scraper is not really built to handle inline javascript before saving HTML currently (it was designed to take screenshot).

Can you share more on how you generate the html with UID for DMR? What tools did you use? 

Thank you!

**Reply by xhluca** [reply]
to get a uid you will need to either (a) inject that at the JS level, (b) artificially add them post-hoc in lxml.

**Reply by matbeedotcom** [reply]
So is the idea that we modify the DOM and attach a (randomly) generated UID to the <body> element? When does the first UID attach, and where?

In reference to this: "what-you-injected-into-elements-on-javascript-side". When a DOM is loaded, at what point is a UID attached, and to what?

**Reply by matbeedotcom** [reply]
If I'm understanding correctly- as we're inferencing an element /won't/ have a UID until we create one after executing `format_relevant_elements_for_single_turn` and then modifying the DOM element for each element_record? Then the model returns the score for each element_record- and in turn we provide it to the LLM. The UID just allows us to select the element by simple string response?

**Reply by xhluca** [reply]
In `format_relevant_elements_for_single_turn` we are already assuming a uid exist (e.g. the attribute would be default named `data-webtasks-id` but that's up to you). When the html is loaded into the `wl.Turn` object, the elements already have a unique id, so it has to be done before. but you can probably do it in `lxml` instead and would probably have the same effect, but when you execute the action you'll need further postprocessing

**Reply by xhluca** [reply]
the uid selection is literal - e.g. `click(uid=1234)` then you are actually finding an element with `<uidkey>=1234` attribute (that can be done in js, selenium, etc) and clicking on it. so it's not an abstraction, it's literally the ~~only way~~ _way we would use_ to select an element during deployment. hence it is recommended to use injection, but if you have a xpath <-> uid mapping and add those uids to the html via lxml in python you won't need to inject, as you can post process the actions predicted by the model by back-mapping uid to xpath and select by xpath instead of selecting attribute.

**Reply by matbeedotcom** [reply]
Okay perfect, I understand exactly what's going on. I appreciate it. Wasnt sure if the model needed UIDs in the markup or not.

**Reply by georgel** [reply]
@xhluca based on your advice, I created a small script that injects the UIDs and save the bbox to JSON in the html. 

I am trying to following your advice above to generate a DMR version of the page. I want to understand the effective of DMR compared to the screenshot + action candidate.  SInce I eventually want to try the DMR result with webllama, should I start by editing eval or processing.py in the DMR directory? or the Llama directory?

If it's possible, can you give me some pointers if I just want to go from HTML to DMR result?

Here's the script I created to inject UID: https://gist.github.com/georgel/7177681813ad4bd213b6156deb1aaf0a

**Reply by xhluca** [reply]
I am working on adding support for running DMR on the fly alongside an action model; however it's a WIP and a bit messy atm, releasing it as is would probably be more confusing than helpful, but I will announce when it will work.

For now, I think that `dmr/processing.py` and `dmr/evaluate.py` for DMR are good start for running it dynamically, then `llama/processing.py` would be used to process the candidate elements output by DMR in `dmr/evaluate.py` (instead of saving as json you just use the dictionary output as is).

**Reply by georgel** [reply]
Looking forward to the DMR on the fly support! Thank you

**Reply by matbeedotcom** [reply]
I have the beginning support for Weblinx and Playwright. Its not perfect and I'm still hacking away at it. Hacking, as in, with a figurative hatchet.
https://github.com/matbee-eth/WeblinxRePlaywrightBrowser

**Reply by tupini07** [reply]
Hi @xhluca , I'm curious if there are any news with on-the-fly DMR? I'm working on creating something like https://github.com/ddupont808/GPT-4V-Act and was really interested in leveraging the code you've already written to get candidates so to minimize the number of options the model has to deal with!

**Reply by xhluca** [reply]
@tupini07 You asked just in time! I just (a few mins ago) included the pre-release for our new experimental API designed to simplify the pipeline for webllama model inference: https://github.com/McGill-NLP/webllama/releases/tag/0.1.0pre1

Since it's experimental, it is not guaranteed to work in a way you expect. In fact, it might fail in subtle ways sometimes... And there's no guarantee it will work well on scenarios that differ from training or test sets. So if you find things to improve, please open an issue & PRs are welcome!


Here's what the code should look like now (much simpler than before), as per the docs/README.md:

```python
import webllama.experimental as wa

# We will initialize our processor, which helps us prepare the input for action model
proc = wa.processing.WebTurnProcessor(tokenizer=act_model.tokenizer)

# Step 1: prepare query, run DMR and prepare retrieved candidates
query_dmr = proc.prepare_dmr_query(action_history, state)
elems = proc.prepare_dmr_elements(state=state)
scores = wa.functions.compute_dmr_scores(dmr, query_dmr, elems)
top_cands, cands_uids = wa.functions.get_top_dmr_candidates(elems, scores, proc.top_k)
cands_str = proc.prepare_candidates(top_cands)

# Step 2: format candidates, utterances, state, and previous actions
html = proc.prepare_state_html(state.html, cands_uids=cands_uids)
utterances = proc.prepare_instructor_chat(action_history, state)
prev_actions = proc.prepare_prev_actions(action_history, state)

# Let's use the default system prompt template, but you can also use your own
sys_prompt_template: str = proc.default_system_prompt_template
sys_prompt = sys_prompt_template.format(
    html=html,
    utterances=utterances,
    candidates=cands_str,
    # ...
)
input_chat = proc.convert_to_chat_list(sys_prompt, prev_actions)

# Use your tokenizer to convert the input to string and pass it to the action model
input_str = act_model.tokenizer.apply_chat_template(input_chat, tokenize=False)
output = act_model(input_str, ...)
pred_action = proc.process_action_model_output(output, state.index, elems)
a = wa.classes.Action.from_dict(pred_action)
```
